packages/@lwc/template-compiler/src/parser-next/index.ts

```diff
@@ -159,6 +163,16 @@ function parseElement(ctx: ParserCtx, elementNode: parse5.Element): IRElement {
     return element;
 }
 
+function parseElementType(parse5Elm: parse5.Element): Element | Component {
```

Comment:
Should we create a new type __ = Element | Component?

Reply:
You can create an alias `type ElementOrComponent = Element | Component`. 

Looking at the PR, this is only needed for retrieving the element tag name. Correct?

Reply:
I'm actually using the `parseElementType` function to determine if the node to be created will be a [Component](https://github.com/salesforce/lwc/blob/ce3fb5b8822ac5dbe71e6165ea7b65f4fbf54c67/packages/%40lwc/template-compiler/src/shared-next/types.ts#L153) or an [Element](https://github.com/salesforce/lwc/blob/ce3fb5b8822ac5dbe71e6165ea7b65f4fbf54c67/packages/%40lwc/template-compiler/src/shared-next/types.ts#L144).  

The logic that looks through the element tag name in `parseElementType` comes from the original [applyComponent](url) function.

What I was going for was something like this:

```
const element: ElementOrComponent = parseElementType(elementNode);
```

The element variable would then contain either an `Element` or a `Component`, which would be passed to the rest of the functions in `parseElement` to complete parsing.

ex:

```
const element: ElementOrComponent = parseElementType(elementNode);
const parsedAttr = parseAttributes(ctx, elementNode);

applyForEach(ctx, element, parsedAttr);
applyIterator(ctx, forEach, element, parsedAttr);
applyIf(ctx, element, parsedAttr);
    
```

My thoughts were that we will have either a `Component` or an `Element` but not both at the same time during the `parseElement` function call but let me know if I'm on the right track.

Reply:
> My thoughts were that we will have either a `Component` or an `Element` but not both at the same time during a single `parseElement` function call but let me know if I'm on the right track.

With the current AST shape, the `Element` AST shape and the `Component` AST shape are disjoint. In practice a component is an element, but not all elements are components. All the components are in the HTML namespace (they can't be SVG or MathML elements).

If it makes your life easier you can for instance make the `Component` interface inherit from the `Element` one.

Reply:
> Also wondering if this is the right way of determining if the node should be a `Component` or `Element`.

Today, LWC treats all the custom elements in the template as LWC components (might change in the future). You can find the logic to distinguish between components and elements in the current `applyComponent` routine.